cmd/juju/user/login.go

```diff
@@ -315,60 +282,109 @@ func (c *loginCommand) publicControllerDetails(host string) (jujuclient.Controll
 	}
 	dialOpts := api.DefaultDialOpts()
 	dialOpts.BakeryClient = bclient
-	conn, err := apiOpen(&c.JujuCommandBase, &api.Info{
-		Addrs: []string{apiAddr},
-	}, dialOpts)
+
+	dial := func(d *jujuclient.AccountDetails) (api.Connection, error) {
+		var tag names.Tag
+		if d.User != "" {
+			tag = names.NewUserTag(d.User)
+		}
+		return apiOpen(&c.JujuCommandBase, &api.Info{
+			Tag:      tag,
+			Password: d.Password,
+			Addrs:    []string{host},
+		}, dialOpts)
+	}
+	conn, accountDetails, err := c.login(ctx, currentAccountDetails, dial)
 	if err != nil {
 		return fail(errors.Trace(err))
 	}
-	defer conn.Close()
-	user, ok := conn.AuthTag().(names.UserTag)
-	if !ok {
-		return fail(errors.Errorf("logged in as %v, not a user", conn.AuthTag()))
-	}
 	// If we get to here, then we have a cached macaroon for the registered
 	// user. If we encounter an error after here, we need to clear it.
 	c.onRunError = func() {
-		if err := c.ClearControllerMacaroons([]string{apiAddr}); err != nil {
+		if err := c.ClearControllerMacaroons([]string{host}); err != nil {
 			logger.Errorf("failed to clear macaroon: %v", err)
 		}
 	}
-	return jujuclient.ControllerDetails{
-			APIEndpoints:   []string{apiAddr},
+	return conn,
+		&jujuclient.ControllerDetails{
+			APIEndpoints:   []string{host},
 			ControllerUUID: conn.ControllerTag().Id(),
-		}, jujuclient.AccountDetails{
-			User:            user.Id(),
-			LastKnownAccess: conn.ControllerAccess(),
-		}, nil
+		}, accountDetails, nil
 }
 
-// updateController updates the controller and account details in the given client store,
-// using the given controller name and adding the controller if necessary.
-func (c *loginCommand) updateController(
-	store jujuclient.ClientStore,
-	controllerName string,
-	controllerDetails jujuclient.ControllerDetails,
-	accountDetails jujuclient.AccountDetails,
-) error {
-	// Check that the same controller isn't already stored, so that we
-	// can avoid needlessly asking for a controller name in that case.
-	all, err := store.AllControllers()
-	if err != nil {
-		return errors.Trace(err)
-	}
-	for name, ctl := range all {
-		if ctl.ControllerUUID == controllerDetails.ControllerUUID {
-			// TODO(rogpeppe) lp#1614010 Succeed but override the account details in this case?
-			return errors.Errorf("controller is already registered as %q", name)
+var errAlreadyLoggedIn = errors.Errorf(`already logged in
```

Comment:
Might it be possible to be clearer to users who are used to the 'controller already registered as' to be more explicit that the user is already logged into the controller in question, and it's available to them? Perhaps just adding the controller name / location to this message?